sealing halt on exec fork

AlexHentschel opened the pull request:
implements https://github.com/dapperlabs/flow-internal/issues/1294

#### Goals
* If sealing has been halted due to an execution fork, this halting condition should be persisted. A consensus node should not continue sealing after reboot.
* encapsulate the logic into a dedicated component
* properly log any inconsistent seals (previously we only printed them to std out) 

The solution in this PR is intended to serve as an additional safety net against verification-sealing bugs: 
* Until we have Missing Collection Challenges and approvals, there should only ever exist a single _correct_ execution result for each block. 
* Up to this point, we can keep the current solution (probably with minor modifications), so sealing will halt even if we have two conflicting but approved execution results. 

### High-level Implementation Overview

I found it an elegant solution to delegate the logic for detecting execution forks to the `IncorporatedResultSeals` mempool. 
* thereby, we don't bloat `consensus.Builder` or `matching.Engine` with this somewhat temporary logic
* to further modularize the code, I have implemented the logic as a **wrapper around the existing `IncorporatedResultSeals` interface**. The wrapper is completely agnostic of the internal implementation

_Minor infrastructure revisions_ contained in this PR
* Some separation in the name spaces of our unittest fixtures: For example, for many fixtures, it would be nice to set the Block the entity pertains to. However, `WithBlock(blockID flow.Identifier)` can only be defined once in unittest. This makes it very hard to determine what options are available for the individual fixtures. Naming becomes very convoluted and lengthy. https://github.com/onflow/flow-go/blob/d61f61f255d5f1d83f84fda1d0abdf5ffe60cf5b/cmd/util/cmd/block_hash_by_height_test.go#L118

AlexHentschel commented on model/flow/incorporated_result.go:
```diff
@@ -35,7 +37,14 @@ func NewIncorporatedResult(incorporatedBlockID Identifier, result *ExecutionResu
 // ID implements flow.Entity.ID for IncorporatedResult to make it capable of
 // being stored directly in mempools and storage.
 func (ir *IncorporatedResult) ID() Identifier {
```
The `IncorporatedResult` is used to store approval signatures when they are received: https://github.com/onflow/flow-go/blob/7c481533a9d41d4c435d2dbb9542e94452c07928/model/flow/incorporated_result.go#L26 When including `aggregatedSignatures` in the ID computation, this changes the ID of the entity, which should not be the case.

AlexHentschel replied:
💡 thanks for pointing this out @arrivets . I was not aware that RLP did skip private fields. 
Obviously, I didn't read the test you wrote: https://github.com/onflow/flow-go/blob/49fb4a12f45253a16441c43ae6da29068abd5d21/model/flow/incorporated_result_test.go#L12-L14 🤦 Apologies. 

Rolled back my changes to `incorporated_result.go`

AlexHentschel commented on module/mempool/stdmap/backend.go:
```diff
@@ -101,18 +101,18 @@ func (b *Backdata) Hash() flow.Identifier {
-	ejectionCallback OnEjection
+	limit             uint
+	eject             EjectFunc
+	ejectionCallbacks []mempool.OnEjection
```
@zhangchiqing I went back to using a list of ejection callbacks. After trying around a bit, I found this solution to be much more general compared to a single callback. Trying to give a high-level summary here:
* some mempools don't store the entities directly in the backend, but use composite structure, e.g.: https://github.com/onflow/flow-go/blob/7c481533a9d41d4c435d2dbb9542e94452c07928/module/mempool/stdmap/incorporated_results.go#L45
* To not expose this internal details to an externally-provided ejection callback, the callback needs to be wrapped before it is injected into the `Backend`. 
* By allowing the backend to serve multiple callbacks:
   - it is trivial for a mempool to wrap any externally-provided callback with the necessary logic before registering the (wrapped) ejection callback with the `Backend`
   - In contrast, when the `Backend` only accepts a single callback, the situation is much more complex for a mempool that has an internal ejection callback as well as accepts external ejection callbacks (like [`IncorporatedResults`](https://github.com/onflow/flow-go/blob/7c481533a9d41d4c435d2dbb9542e94452c07928/module/mempool/stdmap/incorporated_results.go)):
    - the mempool must implement a multiplexer which serves its internal mempool as well as an external mempool

To me, it seemed much more straightforward, to include the functionality to serve multiple callbacks into the Backend, specifically so as it is only adds two lines of code here. If you have other ideas, lets sync Leo. I tried for quite some time and this seemed to be by far the easiest (and least error-prone) solution.

zhangchiqing commented on module/mempool/consensus/exec_fork_suppressor.go:
```diff
@@ -0,0 +1,345 @@
+		if err != nil {
+			return fmt.Errorf("failed to update execution-fork-detected flag: %w", err)
+		}
+		return ExecutionForkErr
```
I was looking for a test case that verifies this error will be raised, but didn't find.

There are different ways to test, the simplest would be a unittest that test this function. Or a test that tests the `Add` function.

AlexHentschel replied:
thanks for flagging this. The `ExecutionForkErr` is only used internally within `ExecForkSuppressor`: 
https://github.com/onflow/flow-go/blob/49fb4a12f45253a16441c43ae6da29068abd5d21/module/mempool/consensus/exec_fork_suppressor.go#L136-L140 I changed the error to be non-exported to avoid confusion. 

For context:
* When implementing `ExecForkSuppressor`, I decided not to error on `Add` in case a fork is detected. Instead, the `ExecForkSuppressor` will internally clear the wrapped mempool and drop any subsequent seals which are added. In my view, this provides the following advantage:
   -  The higher-level business logic can be oblivious to the existence of the `ExecForkSuppressor` wrapper. Otherwise, we would have to include additional logic to treat the specific errors from the wrapper (which is only semi-temporary). 
   -  Not erroring allows the higher-level business logic to potentially continue (just with halted sealing)

AlexHentschel commented on module/mempool/consensus/exec_fork_suppressor.go:
```diff
@@ -0,0 +1,345 @@
+}
+
+// readExecutionForkDetectedFlag attempts to read the flag ExecutionForkDetected from the database.
+// In case no value is persisted, the default value is written to the database.
```
👍 implemented

AlexHentschel commented on module/mempool/consensus/exec_fork_suppressor.go:
```diff
@@ -0,0 +1,345 @@
+		log.Error().Msg("inconsistent seals for the same block")
+		s.seals.Clear()
+		s.execForkDetected = true
+		err := operation.RetryOnConflict(s.db.Update, operation.UpdateExecutionForkDetected(true))
```
> storing the "evidence"
👍 implemented

> if the key exists [...] crash the node.
👍 implemented. For context:
* what we do in case we detect conflicting seals is now encapsulated in https://github.com/onflow/flow-go/blob/33de9d9fc417ea0319aabe3e35ed6e461d9d070d/module/mempool/consensus/exec_fork_actor.go#L12-L14 This allows us to easily replace the logic (and potentially keep going, just without sealing)

AlexHentschel commented on module/builder/consensus/builder_test.go:
```diff
@@ -543,6 +542,54 @@ func (bs *BuilderSuite) TestPayloadSealSomeValid() {
 	bs.Assert().ElementsMatch(bs.chain, bs.assembled.Seals, "should have included only valid chain of seals")
 }
 
+// TestPayloadSealSomeValidOnFork verifies that builder only includes seals whose
```
thanks for noticing. Fixed the test's doc.

AlexHentschel commented on module/builder/consensus/builder_test.go:
```diff
@@ -524,15 +515,23 @@ func (bs *BuilderSuite) TestPayloadSealAllValid() {
 }
 
-func (bs *BuilderSuite) TestPayloadSealSomeValid() {
+// TestPayloadSealSomeValidOnFork verifies that builder only includes seals whose
```
Sorry, this was an outdated artifact from work-in-progress. Removed the artifact.

AlexHentschel commented on module/builder/consensus/builder_test.go:
```diff
@@ -543,6 +542,54 @@ func (bs *BuilderSuite) TestPayloadSealSomeValid() {
+	//bs.T().Skip()
+
+	// in the test setup, we already create a single fork
+	//  [first] <- [F0] <- [F1] <- [F2] <- [F3] <- [A0] <- [A1] <- [A2] <- [A3]
+	// Where block
+	//   * [first] is sealed and finalized
+	//   * [F0] ... [F3] are finalized but _not_ sealed
+	//   * [A0] ... [A3] are _not_ finalized and _not_ sealed
+	// We now create an additional fork:  [F3] <- [B0] <- [B1] <- ... <- [B7]
```
thanks for the reference. The test for the ` Builder` already contained its own implementation: https://github.com/onflow/flow-go/blob/addcb4b127d1b9b7dca41dc5f3c03e36a69da2fd/module/builder/consensus/builder_test.go#L87-L92
it has some specialized functionality for testing the  payload `Builder`:
* creates execution receipts and seals for the blocks and includes them into the block's payloads
* tracks the created seals and blocks in the mempool mocks

I am inclined to keep the current test structure as I just added a test. 😅

AlexHentschel commented:
Thanks for the insightful review comments @zhangchiqing . I have implemented most of your suggestions. 

Please note that I have split up the logic a bit more: 
* the  `ExecForkSuppressor` will enforce that sealing halts when an execution fork is detected 
* whether the consensus node crashes or not is now encapsulated in a separate function: https://github.com/onflow/flow-go/blob/33de9d9fc417ea0319aabe3e35ed6e461d9d070d/module/mempool/consensus/exec_fork_actor.go#L12-L14 
   - This allows us to easily change whether or not we want to crash the consensus node or keep going (with halted sealing) 
   - Though, based on our sync today, I only implemented crashing the consensus node

zhangchiqing commented on module/mempool/consensus/exec_fork_suppressor.go:
```diff
@@ -0,0 +1,353 @@
+type ExecForkSuppressor struct {
+	mutex            sync.RWMutex
+	seals            mempool.IncorporatedResultSeals
+	sealsForBlock    map[flow.Identifier]sealSet // map BlockID -> set of IncorporatedResultSeal
+	execForkDetected bool
+	onExecFork       ExecForkActor
```
Optional suggestion:

If we are committing to the approach that we will crash as soon as we detect forks, then we could simply the logic here. 

For instance, we could use the default log and crash as the `onExecFork`, and `execForkDetected` is not need, because we will crash if detected, and if we haven't crashed, then it means we haven't detected execFork yet. And we don't need `sealSet` either, because there won't be 3 conflicting seals for the same block. As soon as there are 2 conflicting, then we will log them, restore them as evidence and crash.

AlexHentschel replied:
Agree, it is complexity we don't utilize at this point and thereby a risk for additional bugs.  Though, the code is quite well tested so I feel this risk is relatively low. I still think the benefit of being able to boot up consensus nodes with only halted sealing but otherwise functioning block production outweighs the small risk.

zhangchiqing reviewed (approved):
Had some optional suggestions. Other than that looks great 👍

AlexHentschel commented on module/builder/consensus/builder_test.go:
```diff
@@ -543,6 +542,54 @@ func (bs *BuilderSuite) TestPayloadSealSomeValid() {
 
+// TestPayloadSealSomeValidOnFork verifies that builder only includes seals whose
+//
+// if an IncorporatedResultSeal is included in mempool, whose final state does not match the execution result
```
added

AlexHentschel commented on module/builder/consensus/builder_test.go:
```diff
@@ -543,6 +542,54 @@ func (bs *BuilderSuite) TestPayloadSealSomeValid() {
+	//  * currently, the builder relies on the IncorporatedResultSeal to be constructed
+	//    correctly by the matching engine
+	//  * with the implementation of this test, the matching engine
+	//bs.T().Skip()
```
thanks. cleaned up test

AlexHentschel commented on module/builder/consensus/builder_test.go:
```diff
@@ -543,6 +542,54 @@ func (bs *BuilderSuite) TestPayloadSealSomeValid() {
+	//   * [A0] ... [A3] are _not_ finalized and _not_ sealed
+	// We now create an additional fork:  [F3] <- [B0] <- [B1] <- ... <- [B7]
+	var forkHead *flow.Block
+	for f := 0; f < 100; f++ {
```
nope. Good point. 
Changed to only add one fork instead of 100.

AlexHentschel commented on module/mempool/consensus/exec_fork_suppressor_test.go:
```diff
@@ -0,0 +1,386 @@
+// Test_Add adds IncorporatedResultSeals for
+//   * 2 different blocks
+//   * for each block, we generate one specific result,
+//     for which we add 4 IncorporatedResultSeals
```
fixed

AlexHentschel commented on module/mempool/consensus/exec_fork_suppressor.go:
```diff
@@ -0,0 +1,345 @@
+	// important to prevent memory leak
+	newSealID := newSeal.ID()
+	if _, exists := s.seals.ByID(newSealID); !exists {
+		return added, nil
```
Totally agree with your thinking and would prefer a different return value in this particular case. Before we decide what the return value should be, lets take a look at the larger picture:
 * `ExecForkSuppressor` is a wrapper a around `mempool.IncorporatedResultSeals`. So how does the implementation behave when you `Add` an entity and the mempool overflows?
   - `mempool.IncorporatedResultSeals` is implemented by   https://github.com/onflow/flow-go/blob/d14a1e7a634d3d458a09406f2bbff96bb0701777/module/mempool/stdmap/incorporated_result_seals.go#L21-L24 which calls into `Backend` https://github.com/onflow/flow-go/blob/d14a1e7a634d3d458a09406f2bbff96bb0701777/module/mempool/stdmap/backend.go#L131-L138 and subsequently into `Backdata` https://github.com/onflow/flow-go/blob/d14a1e7a634d3d458a09406f2bbff96bb0701777/module/mempool/stdmap/backend.go#L31-L40
   - Specifically, the `Backend` always stores the element in `Backdata`. The **return value indicates**
      - `false`: entity is already present in the mempool and was deduplicated 
      - `true`: entity is new from the mempool's perspective and can be added (ignoring mempool's capacity limitations) 
   - The return value makes **no assertion about whether or not an element is subsequently ejected** 
* **Essentially, `IncorporatedResultSeals.Add(seal)` returns `true` if the entity is addable. Even if the mempool overflows and decides to eject the  `seal` right away, the seal was still "addable".** Note that this convention applies to all mempools backed by `Backend`. 

**`ExecForkSuppressor` is only a wrapper, and higher-level business logic should be agnostic to whether or not the wrapper exists. Therefore, I concluded that the wrapper should implement the same convention as the wrapped mempool.**

Nevertheless, @arrivets you are raising a valid point: 
* Given that a mempool has the ability to eject elements, can we base some algorithmic decisions on the return value? 
* If the mempool returns `false`, it is telling us that the entity is already stored. However, this might change any moment, as another routine might concurrently add another entity leading to the ejection of our entity. 
* Likewise, if the mempool returns `true`, the entity might be ejected right away as well.

AlexHentschel commented on cmd/consensus/main.go:
```diff
@@ -131,7 +132,11 @@ func main() {
 			// the chain of seals
 			ejector := ejectors.NewLatestIncorporatedResultSeal(node.Storage.Headers)
-			seals = stdmap.NewIncorporatedResultSeals(sealLimit, stdmap.WithEject(ejector.Eject))
+			seals = stdmap.NewIncorporatedResultSeals(stdmap.WithLimit(sealLimit), stdmap.WithEject(ejector.Eject))
+			seals, err = consensusMempools.NewExecStateForkSuppressor(consensusMempools.LogForkAndCrash(node.Logger), seals, node.DB, node.Logger)
```
done

AlexHentschel commented on module/mempool/consensus/exec_fork_suppressor.go:
```diff
@@ -0,0 +1,353 @@
+	result := irSeal.IncorporatedResult.Result
+
+	initialState, ok := result.InitialStateCommit()
+	if !ok || len(initialState) < 1 {
```
done

AlexHentschel commented on module/mempool/consensus/exec_fork_suppressor.go:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	finalState, ok := result.FinalStateCommitment()
+	if !ok || len(finalState) < 1 {
```
done

AlexHentschel commented on module/mempool/consensus/exec_fork_suppressor.go:
```diff
@@ -0,0 +1,353 @@
+
+// All returns all the IncorporatedResultSeals in the mempool
+func (s *ExecForkSuppressor) All() []*flow.IncorporatedResultSeal {
+	s.mutex.RLock()
```
Interesting suggestion. 🤔 
You are probably right: we don't need to lock for methods that only do an atomic read from the wrapped mempool. But I am not 100% sure. 

* We certainly need to lock `ExecForkSuppressor` for modifying operations such as `Add` or `Rem` as they interact with the wrapped mempool multiple times. 
* So the question is then: can anything bad happen if we read from the wrapped mempool, while  `Add` or `Rem` are modifying it at the same time? 

As long as I am not sure, I am hesitant to remove the lock here. With the lock, we are sure that we don't have any concurrency edge cases.
